src/main/java/org/openRealmOfStars/utilities/DataSources.java

```diff
@@ -0,0 +1,126 @@
+package org.openRealmOfStars.utilities;
+/*
+ * Open Realm of Stars game project
+ * Copyright (C) 2023 BottledByte
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see http://www.gnu.org/licenses/
+ */
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Class for looking-up files in various data sources.
+ *
+ * <p>
+ * Intended to allow (rather basic) modding.
+ * Game data can be "shadowed" by data with same name
+ * in other source, which has higher priority.
+ * In case the data is not found in external source,
+ * it will fallback to trying to load it from game JAR.
+ * </p>
+ *
+ * Only alternative external data source is currenly
+ * <i>"gamedata" directory, in directory with game JAR</i>.
+ */
+public final class DataSources {
+  /** The Singleton */
+  private static final DataSources SINGLETON = new DataSources();
+
+  /**
+   * Search for requested file with relativeName in all sources,
+   * prefer external, fallback to JAR.
+   * @param relativeName data "name"
+   * @return URL to data or empty
+   */
+  public static Optional<URL> getDataUrl(final String relativeName) {
+    return SINGLETON.getUrl(relativeName);
+  }
+
+  /** List of paths to try before fallback */
+  private ArrayList<String> shadowPaths;
+
+  /** Construct new object */
+  private DataSources() {
+    this.shadowPaths = new ArrayList<>();
+
+    String srcJarPath = null;
+    try {
+      srcJarPath = DataSources.class.getProtectionDomain()
+          .getCodeSource().getLocation().toURI().getPath();
+    } catch (URISyntaxException e) {
```

Comment:
Empty catch block do not pass through build. Put debug error logging here for example. I am not sure how likely URISyntaxException is.